### ISSUE TITLE
Network jitter extreme until you "alt-tab" out of Moonlight on Steam Deck

### ISSUE THREAD

**Issue by ap3nr0ts** [moonlight-stream/moonlight-qt#1152]
**Describe the bug**
Basically, the problem is simple and very easy to solve (every time it happens again). Any time I stream via Moonlight on my Steam Deck (OLED, 5GHz, PC is cabled, SD right next to router), the stream becomes very choppy after a certain amount of time - let's say 10 minutes - until I go out of the stream, open something such as settings, and return to the stream. Then it becomes good again and stays good. No more jitter. It's not a huge issue, but I'd like to resolve it if possible.

**Steps to reproduce**
Just start the stream. I use 1280x800 met V-sync and frame pacing. 25 is my bitrate (upped it just a bit for any possible quality increase). My internet on the Steam Deck gives me around 500Mbps down and up, however. I've also made my router a bit more stable by only allowing 40Hz channels (otherwise, speed can be up to 1Gbps). The channel used is usually quiet/me by myself.

**Affected games**
Basically any game does this.

**Moonlight settings (please complete the following information)**
Other settings that might be relevant: I use hardware decoding and the stream defaults to HEVC. Wouldn't know why this would be a problem, however, since the issue resolves after that aforementioned "alt-tab".

**Server PC details (please complete the following information)**
 - OS: Windows 11
 - Sunlight latest version
 - Nvidia GPU driver latest version
 - Windows Firewall

**Comment by absamue** [comment]
@ap3nr0ts 

Are you using Moonlight directly or utilizing MoonDeck? I have been having this issue as well and have been trying to isolate whether this is related to Moonlight, MoonDeck, SteamOS, or the wifi hardware itself.

**Comment by ap3nr0ts** [comment]
@absamue I have it on Moonlight. First, I used the flatpak. Now, I'm using the AppImage. I thought it was gone, but that only lasted for 1 or 2 sessions, haha. (Read somewhere that AppImage is maybe more stable.) No MoonDeck here.

I also think it's something Steam Deck-related, since it disappears immediately after opening the list of wifi options and going back to the game. That's basically my solution right now: pause the game, go to wifi settings, return to the game, and the jitter goes down from 25% or more to 0% again.

Again, it's not a big issue, and only happens one time every session. But if it can be fixed, why not. You've got exactly the same thing? Got an OLED as well? Then that wifi chip could be the blame. Don't remember this happening on my LCD.

**Comment by absamue** [comment]
Yep I am on the OLED model, didn't have the LCD for comparison.

I can't seem to get the issue to consistently happen, but it seems to show up most commonly when I am wifi streaming Risk of Rain 2. Will try your workaround, as I was having to sleep-wake and reconnect the session.

**Comment by ap3nr0ts** [comment]
Then I don't know if you have my problem, haha. I have never seen it come back after the first try. Only if I totally reboot the stream. It's a one-off thing, which is why I find it so strange (and not that troublesome).

**Comment by HauieT** [comment]
Have this exact same problem. Going to bluetooth menu then go back to game seems to also cure this problem. I think its a steamdeck issue unfortunately, and valve is not doing anything to fix it ;(

**Comment by ap3nr0ts** [comment]
> Have this exact same problem. Going to bluetooth menu then go back to game seems to also cure this problem. Another thing i’d like to add is that, whenever these lags occurs, I can observe from my router that the tx rate with the steam deck hits rock bottom for some reason, like it is going in power save mode or something. Overall, my wifi works normally.

I'll try that menu tonight. My guess is that it's a Steam Deck hard- or software thing as well then. Maybe the wifi chip is going into a lower power mode at first, and needs to get a good kick in the behind. (Just an assumption based on the fact wifi was having hiccups after sleep mode, if I remember correctly).

**Comment by HauieT** [comment]
> > Have this exact same problem. Going to bluetooth menu then go back to game seems to also cure this problem. Another thing i’d like to add is that, whenever these lags occurs, I can observe from my router that the tx rate with the steam deck hits rock bottom for some reason, like it is going in power save mode or something. Overall, my wifi works normally.
> 
> I'll try that menu tonight. My guess is that it's a Steam Deck hard- or software thing as well then. Maybe the wifi chip is going into a lower power mode at first, and needs to get a good kick in the behind. (Just an assumption based on the fact wifi was having hiccups after sleep mode, if I remember correctly.)

Yeah this is my theory too. For some reason when waking steamdeck up, the wifi adapter goes into power save mode within 10-20min of use with moonlight (I could not observe that this happens when you just play games straight from steamdeck, or just stay in steam menu.
Once it goes into power save, you can wake by going into bluetooth menu/wifi menu (since they are on the same adapter), it forces the adapter to go full power again for the rest of the session until the deck is restarted.
Super weird behavior.

**Comment by ap3nr0ts** [comment]
> > > Have this exact same problem. Going to bluetooth menu then go back to game seems to also cure this problem. Another thing i’d like to add is that, whenever these lags occurs, I can observe from my router that the tx rate with the steam deck hits rock bottom for some reason, like it is going in power save mode or something. Overall, my wifi works normally.
> > 
> > 
> > I'll try that menu tonight. My guess is that it's a Steam Deck hard- or software thing as well then. Maybe the wifi chip is going into a lower power mode at first, and needs to get a good kick in the behind. (Just an assumption based on the fact wifi was having hiccups after sleep mode, if I remember correctly.)
> 
> Yeah this is my theory too. For some reason when waking steamdeck up, the wifi adapter goes into power save mode within 10-20min of use with moonlight (I could not observe that this happens when you just play games straight from steamdeck, or just stay in steam menu. Once it goes into power save, you can wake by going into bluetooth menu/wifi menu (since they are on the same adapter), it forces the adapter to go full power again for the rest of the session until the deck is restarted. Super weird behavior.

Maybe the following is an interesting try then: "To turn this off, you need to activate developer mode in the deck's settings. Then there will be a new entry on the left side panel at the bottom for developer. Go there, scroll down and find wifi power management and turn it off. It'll ask you to reboot. Done." Found this on reddit about wifi power management.

**Comment by Herro95** [comment]
> > > > Have this exact same problem. Going to bluetooth menu then go back to game seems to also cure this problem. Another thing i’d like to add is that, whenever these lags occurs, I can observe from my router that the tx rate with the steam deck hits rock bottom for some reason, like it is going in power save mode or something. Overall, my wifi works normally.
> > > 
> > > 
> > > I'll try that menu tonight. My guess is that it's a Steam Deck hard- or software thing as well then. Maybe the wifi chip is going into a lower power mode at first, and needs to get a good kick in the behind. (Just an assumption based on the fact wifi was having hiccups after sleep mode, if I remember correctly.)
> > 
> > 
> > Yeah this is my theory too. For some reason when waking steamdeck up, the wifi adapter goes into power save mode within 10-20min of use with moonlight (I could not observe that this happens when you just play games straight from steamdeck, or just stay in steam menu. Once it goes into power save, you can wake by going into bluetooth menu/wifi menu (since they are on the same adapter), it forces the adapter to go full power again for the rest of the session until the deck is restarted. Super weird behavior.
> 
> Maybe the following is an interesting try then: "To turn this off, you need to activate developer mode in the deck's settings. Then there will be a new entry on the left side panel at the bottom for developer. Go there, scroll down and find wifi power management and turn it off. It'll ask you to reboot. Done." Found this on reddit about wifi power management.

We all did this weeks ago and it doesn‘t solve the power saving problem.. i think it‘s a driver problem or moonlight bug

**Comment by ap3nr0ts** [comment]
@Herro95 Got a link? I've disabled it and didn't have it in 2 games after. "n=1" since it's one night, but I'll try again tomorrow night.

**Comment by Herro95** [comment]
> @Herro95 Got a link? I've disabled it and didn't have it in 2 games after. "n=1" since it's one night, but I'll try again tomorrow night.

https://github.com/ValveSoftware/SteamOS/issues/1256

**Comment by jamespost** [comment]
Just wanted to add that I am experiencing the same issues of network jitter and bad frame-pacing after a few minutes of play every Moonlight session, also on an OLED Deck. One thing I've noticed is that when streaming games with the built-in Steam Remote Play the jitter never occurs. Have others also noticed this? To me this seems to point to the problem being something to do with Moonlight, not necessarily the Deck's WiFi chip inherently?

**Comment by ap3nr0ts** [comment]
@jamespost @Herro95 Good point. I did get it yesterday, with power management turned off. So you were right, Herro. I could still fix it by opening the menu on the right and quickly going to the wifi settings menu. Which still seems like a weird way to fix something. :P

**Comment by HauieT** [comment]
Have you guys had any success tinkering with this? Tried with flatpak and appimage version of moonlight, no change.

I have also installed merlinwrt on my Asus rt-ax86u pro with no success either.

**Comment by ap3nr0ts** [comment]
@HauieT Nah. No change. I really think it is either SteamOS or Moonlight at play here, since it _always_ gets fixed for me by just opening the wifi menu. Which isn't even a real hassle by navigating via the pop-up menu with the three dots button.

**Comment by HauieT** [comment]
> @HauieT Nah. No change. I really think it is either SteamOS or Moonlight at play here, since it _always_ gets fixed for me by just opening the wifi menu. Which isn't even a real hassle by navigating via the pop-up menu with the three dots button.

Okay, I just ordered a different router to see if that fixes the issue for me. There have been reports of people having no issues as well, so worth a try I guess. 
What router are u using?

**Comment by Herro95** [comment]
> > @HauieT Nah. No change. I really think it is either SteamOS or Moonlight at play here, since it _always_ gets fixed for me by just opening the wifi menu. Which isn't even a real hassle by navigating via the pop-up menu with the three dots button.
> 
> Okay, I just ordered a different router to see if that fixes the issue for me. There have been reports of people having no issues as well, so worth a try I guess. What router are u using?

I already ordered a new router and the problem still exist. I‘m using the AVM Fritz!Box 6690 Wifi 6 Router

**Comment by HauieT** [comment]
> > > @HauieT Nah. No change. I really think it is either SteamOS or Moonlight at play here, since it _always_ gets fixed for me by just opening the wifi menu. Which isn't even a real hassle by navigating via the pop-up menu with the three dots button.
> > 
> > 
> > Okay, I just ordered a different router to see if that fixes the issue for me. There have been reports of people having no issues as well, so worth a try I guess. What router are u using?
> 
> I already ordered a new router and the problem still exist. I‘m using the AVM Fritz!Box 6690 Wifi 6 Router

I see, and what router did you use previously? I have tried 2 different asus routers, and had the same problem on both.
One being wifi 5 and the other wifi 6.
The new router i've ordered is a TP-Link with Wifi 6E, we'll see if that does any difference.

**Comment by Herro95** [comment]
> > > > @HauieT Nah. No change. I really think it is either SteamOS or Moonlight at play here, since it _always_ gets fixed for me by just opening the wifi menu. Which isn't even a real hassle by navigating via the pop-up menu with the three dots button.
> > > 
> > > 
> > > Okay, I just ordered a different router to see if that fixes the issue for me. There have been reports of people having no issues as well, so worth a try I guess. What router are u using?
> > 
> > 
> > I already ordered a new router and the problem still exist. I‘m using the AVM Fritz!Box 6690 Wifi 6 Router
> 
> I see, and what router did you use previously? I have tried 2 different asus routers, and had the same problem on both. One being wifi 5 and the other wifi 6. The new router i've ordered is a TP-Link with Wifi 6E, we'll see if that does any difference.

The AVM Fritz!Box 6591 Wifi 5 Router. Maybe your Wifi6E is working better.

**Comment by HauieT** [comment]
> > > > > @HauieT Nah. No change. I really think it is either SteamOS or Moonlight at play here, since it _always_ gets fixed for me by just opening the wifi menu. Which isn't even a real hassle by navigating via the pop-up menu with the three dots button.
> > > > 
> > > > 
> > > > Okay, I just ordered a different router to see if that fixes the issue for me. There have been reports of people having no issues as well, so worth a try I guess. What router are u using?
> > > 
> > > 
> > > I already ordered a new router and the problem still exist. I‘m using the AVM Fritz!Box 6690 Wifi 6 Router
> > 
> > 
> > I see, and what router did you use previously? I have tried 2 different asus routers, and had the same problem on both. One being wifi 5 and the other wifi 6. The new router i've ordered is a TP-Link with Wifi 6E, we'll see if that does any difference.
> 
> The AVM Fritz!Box 6591 Wifi 5 Router. Maybe your Wifi6E is working better.

ill let you know if I got any luck with it once it arrives and I've had some time testing it out.

**Comment by Herro95** [comment]
Perhaps someone can write a plugin for Decky so that, for example, the WLAN card is addressed every 5 minutes in the background so that this bug does not occur. You can certainly call up the wireless settings with a command line or search for WLAN networks. Maybe that works..

**Comment by Black-Seraph** [comment]
This is entirely a Steam Deck OLED issue.

I have access to both makes of the Deck, LCD & OLED, and only the OLED displays these symptoms.

Furthermore, this isn't a Moonlight-only issue... whenever the Deck decides to do this it also affects game downloads from Steam, cloud syncs, etc.

The throughput pretty much hits the low 20-100kb/s whenever this happens.

I personally see no reason to keep this issue open, at least not on the Moonlight side of things, as this is something that should be fixed by Valve on their side, not worked around in some hasty way in Moonlight.

I also very much believe this to be a WiFi driver issue more than anything else as the Windows WiFi driver (taken from a Lenovo notebook) very much functions properly.

**Comment by HauieT** [comment]
Hey, so Ive had some time with my new router with wifi 6e. The issue seems to have been resolved for me after changing routers, no more random lag warning from moonlight after 10-20min use.

**Comment by HauieT** [comment]
> @Herro95 @Black-Seraph @HauieT Based on everything here, I think Black-Seraph is right. If he can confirm this also happens when downloading etc., it needs to be fixed on Steam's end. What's the best way to reach Valve's developers? A support ticket?
> 
> Edit: created a ticket and a topic. https://steamcommunity.com/app/1675200/discussions/1/4148446223175260491/

People have been flagging valve about this issue since early december. Valve have been very silent about the status unfortunately, so we don’t even know if they are planning a fix it.

**Comment by ap3nr0ts** [comment]
Would a Decky plugin just loading up the wifi menu be a workaround then? Don't know how you'd do that, since opening the menu once is enough and you get great slowdown while in that menu.

**Comment by ap3nr0ts** [comment]
@HauieT @Black-Seraph @Herro95 I've got contact with "sdsupport" now who are letting me run all sorts of tests. Could you tell me the exact circumstances under which you get this error? Also via Steam Remote Play and/or downloading games? That could be useful. :)

**Comment by Swaggerzhan** [comment]
i use ssh connect to deck, and just run powertop while i streaming, looks can temporary fix

**Comment by TinyHuman99** [comment]
> i use ssh connect to deck, and just run powertop while i streaming, looks can temporary fix

Can you elaborate on this? i'm still having the same issue on OLED deck. Thanks

**Comment by Swaggerzhan** [comment]
> > i use ssh connect to deck, and just run powertop while i streaming, looks can temporary fix
> 
> Can you elaborate on this? i'm still having the same issue on OLED deck. Thanks
@TinyHuman99 
just turn deck into desktop mode, use terminal to enable ssh connection by:
systemctl start sshd
systemctl enable sshd
and use another device ssh connect the deck, run the command "sudo powertop", and left it running without closing the ssh connection. and back to using moonlight, and it seems to have fixed the issue...

you can try it; i'm not sure if this method will fix it for you as well, but works on my  OLED deck.

**Comment by DieHertz** [comment]
I have extreme network jitter as well, but I'm on wired 1gbit and alt-tabbing does not help

**Comment by sp0tter-sd** [comment]
Guys, has anyone managed to solve this problem? I’m still facing it. Every time I play for the first time using Moonlight, the connection has a sudden lag spike and freezes. This always happens between 10 and 20 minutes in. The issue is fixed simply by pressing the hotkey to go to the Moonlight menu and then returning to the game. After this first time, the problem doesn’t happen anymore and everything stays stable for hours, even... I tried using the Decky Wi-Fi Locker app as suggested, but it didn’t work for me. I’m using the Steam Deck OLED 512GB. I appreciate any help you can give!

**Comment by Fir3hazard** [comment]
Has anyone ever tried using some kind of "dummy" app/game running in the background to avoid triggering the wifi power saving mode that occurs when streaming?

**Comment by Fir3hazard** [comment]
Latest Steam Deck beta supposedly fixes the SD OLED wifi regression issue. Hoping to see people's results here 🙏

**Comment by malmaud** [comment]
No, this is not fixed in the latest OS. See this comment from a Valve engineer confirming the issue still persists on the SD OLED https://github.com/ValveSoftware/SteamOS/issues/1445#issuecomment-2988732143 and various user reports backing that up.

**Comment by GeorgyDesign** [comment]
> I have extreme network jitter as well, but I'm on wired 1gbit and alt-tabbing does not help
> 
> **So I have found the reason of my issue**: the game recording feature is seemingly using the same encoding/decoding circuitry and causes resource contention. Disabling game recording entirely I now have 0% frames dropped due to "network jitter". Disabling recording selectively for Moonlight has helped, but some <5% jitter still remained. With it enabled I had 33% on average

Steam Recording feature on your PC/laptop or Steam Deck?

> No, this is not fixed in the latest OS. See this comment from a Valve engineer confirming the issue still persists on the SD OLED [ValveSoftware/SteamOS#1445 (comment)](https://github.com/ValveSoftware/SteamOS/issues/1445#issuecomment-2988732143) and various user reports backing that up.

Dang, I’ve enabled “Force WPA Supplicant WiFi backend” option in Developer Mode on my Steam Deck OLED — hopefully it helps, but if not… well, I do hope Valve will fix this issue, because my router and laptop are fine, I’m sure.

**Comment by DieHertz** [comment]
> Steam Recording feature on your PC/laptop or Steam Deck?

On the Deck

**Comment by sladkiybublik** [comment]
It's been almost two years since this problem. I'm shocked. I can't play normally through Moonlight. Sad..

**Comment by br3nn** [comment]
I can’t believe I found your post—I honestly thought I was going mad lol.

I’m on an OLED Steam Deck, with my PC cabled into an ASUS TUF AX4200 (getting those 1Gb link speeds), and the Deck on Wi-Fi. Whether I’m using Steam Big Picture Mode (Middle icon)  or a virtual display for those custom resolutions, I’m hitting the same wall. I’ve tested everything from 35mbps to 60mbps at both 60 and 90 FPS.

Like clockwork, about 5 minutes into Hogwarts Legacy, the red text pops up and the lag kicks in. My "fix" is weirdly simple: I open the Decky/Quick Access menu, hit the Wi-Fi icon (even though I’m already connected), then just jump back into Moonlight. Five seconds later, BOOM—it’s perfect for the rest of the session. I played for two hours straight yesterday without another hitch after doing that.

It feels like the Deck is hunting for channels or dipping into a power-save mode until I "wake it up" by opening the Wi-Fi settings. Since my Logitech G Cloud and iPhone/Backbone never had this issue, I’m convinced it’s a SteamOS/Wi-Fi driver quirk and not my network. Subbed to the thread to see if anyone finds a permanent fix!

**Comment by justfalter** [comment]
I think I have made some progress with sussing out the issue:

1. When my OLED steamdeck is having high latency/jitter, the adapter shows a low receive (rx) MCS index / bitrate relative to the transmission (tx) side of things.
2. Going to the settings page of the steamdeck does resolve the latency/jitter issues, and I see the receive MCS index / bitrate for the adapter rise to the tx levels.
3. Another apparent "fix" is to initiate a wifi scan on the adapter with `iwctl station wlan0 scan`. 

----

I've run into this with my OLED steamdeck.  Like others, going to the steam settings page suddenly jerks things back to life. 

I noticed that, when the jitter starts, the `RxMCS` for wlan0 drops really low (to 0 or 1), while the `TxMCS` remains high (>=7). The wifi MCS (modulation coding scheme) index is a quality of connection indicator for the adapter's Tx and Rx channels.

```
$ iwctl station wlan0 show
--------------------------------------------------------------------------------      
  Settable  Property              Value                                               
--------------------------------------------------------------------------------      
            Scanning              no                                                                                  
            State                 connected                                                                           
            Connected network     xxxxxxx                                                                           
            IPv4 address          192.168.55.245                                      
            ConnectedBss          ea:38:83:xx:xx:xx                                  
            Frequency             5745                                                                                
            Channel               149                                                                                 
            Security              WPA2-Personal + FT                                                                  
            RSSI                  -54 dBm                                                                             
            AverageRSSI           -54 dBm                                             
            RxMode                802.11ax                                                                            
            RxMCS                 0                                                                                   
            TxMode                802.11ax                                                                            
            TxMCS                 8                                                                                   
            TxBitrate             864800 Kbit/s                                       
            RxBitrate             34400 Kbit/s                                                                        
                                                                                                                      
```

It makes sense that network latency would spike when the adapter's effective receive rate is severely hampered.

Like others, loading the steam settings page resolves the issue. When resolved, the `RxMCS` and `RxBitrate` rise to levels comparable to the tx side of things. Wondering about _why_ loading the steam settings menu might appear to fix things -- I figured the only thing having to do with wifi on page load was to list or scan for nearby wifi networks.

Retrieving a list a wifi networks does *not* seem to change anything: `iwctl station wlan0 get-networks`, as the `RxMCS` remained 1 or 0. I'm guessing that this is probably just a list maintained in memory by the iwd daemon.

However, starting a scan for wifi networks _did_ seem to boost the `RxMCS` and `RxBitrate` up: `iwctl station wlan0 scan`

Viewing the adapter's station status afterward showed that the `RxMCS` and `RxBitrate` rose to levels on par with the tx side of things:

```
--------------------------------------------------------------------------------
  Settable  Property              Value                                          
--------------------------------------------------------------------------------
            Scanning              no                                               
            State                 connected                                        
            Connected network     xxxxxxxx                                        
            IPv4 address          192.168.55.245                                   
            ConnectedBss          ea:38:83:xx:xx:xx                       
            Frequency             5745                                             
            Channel               149                                              
            Security              WPA2-Personal + FT                               
            RSSI                  -63 dBm                                          
            AverageRSSI           -64 dBm                                          
            RxMode                802.11ax                                         
            RxMCS                 10                                               
            TxMode                802.11ax                                         
            TxMCS                 9                                                
            TxBitrate             960700 Kbit/s                                    
            RxBitrate             1080600 Kbit/s   
```

**Comment by andygrundman** [comment]
This is good info, but I don't know that this bug is going to do much good here, it should be brought to Valve's attention, if they don't already have an open bug for it.

**Comment by SirvanD** [comment]
In my case, the solution in below project worked, testing for about 5 hours haven't had any connectivity issue while using Moonlight.
Will need more testing of course. Just putting it out there in case anyone wanted to try if you exhausted all other workarounds and only restarting Wifi worked.
https://github.com/doughty247/hifi-wifi
